### PR TITLE
fix: finalize mHC composite metrics on substep end

### PR DESCRIPTION
**Pull request by yuweivvv**
### PR types

Others

### PR changes

Integrate Internal Medicine monitor finalization with optimizer and accumulation callbacks.

### Description

This PR adds callback integration for Internal Medicine monitors:

- Passes the current AMP `GradScaler` to `finalize_scaled_grad_metrics()` before the optimizer step, allowing autograd-hook gradient metrics to be restored to unscaled values;
- Calls `finalize_composite_microbatch()` at `on_substep_end`, preventing mHC composite snapshots from being overwritten during gradient accumulation;
- Uses capability detection to preserve compatibility with monitors that do not implement these optional methods.

The change only affects metric finalization and does not modify model gradients, optimizer behavior, or training semantics.

### CODE REVIEW

File: paddleformers/trainer/trainer_callback.py [hunk 1]
```diff
@@ -1054,6 +1054,15 @@ def on_optimizer_begin(self, args, state, control, scaler=None, **kwargs):
             if finalize is not None:
                 finalize(scaler)
 
+    def on_substep_end(self, args, state, control, **kwargs):
+        if not self._setup_done:
+            return
+
+        for monitor in self._monitor_dict.values():
+            finalize = getattr(monitor, "finalize_composite_microbatch", None)
```
Comment (risemeup1111) [inline paddleformers/trainer/trainer_callback.py:1062]: P1 当前 pinned 的 `llm-internal-medicine` 子模块为 `2d24c2d`，其 PaddleFleet `setup_monitors` 的注册表只有 `qk_stats`、`moe_health`、`massive_act`，并没有 `mhc_health`/`finalize_composite_microbatch`；因此这里的 capability 检测始终找不到目标 monitor，mHC 快照仍会在梯度累积期间被后续 micro-batch 覆盖。请同步升级该子模块（或在本仓库提供对应 monitor 实现）并补充精确依赖下的累积训练测试，确保该回调实际能触发。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [trainer_callback.py:1062](https://github.com/PaddlePaddle/PaddleFormers/blob/e04b123221ef43f22844c4ddd128649a2dbb9b6c/paddleformers/trainer/trainer_callback.py#L1062) | P1 | 🚧 |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by zjjlivein — APPROVED** [review]
